src/abstract-interpretation/data-frame/abstract-interpretation.ts

```diff
@@ -0,0 +1,158 @@
+		let oldDomain = new Map<NodeId, DataFrameDomain>();
+		let newDomain = inputDomain;
+
+		const entryNode: RNode<ParentInformation & AbstractInterpretationInfo> | undefined = dfg.idMap?.get(vertex.id);
+		let node: RNode<AbstractInterpretationInfo> | undefined;
+
+		if(entryNode !== undefined && isRSingleNode(entryNode)) {
+			oldDomain = entryNode.info.dataFrame?.domain ?? oldDomain;
+			newDomain = processDataFrameLeaf(entryNode, new Map(inputDomain), dfg);
+			node = entryNode;
+		} else if(vertex.type === CfgVertexType.EndMarker) {
+			const exitId = getNodeIdForExitVertex(vertex.id);
+			const exitNode: RNode<ParentInformation & AbstractInterpretationInfo> | undefined = exitId !== undefined ? dfg.idMap?.get(exitId) : undefined;
+
+			if(exitNode !== undefined && !isRSingleNode(exitNode)) {
+				oldDomain = exitNode.info.dataFrame?.domain ?? oldDomain;
+				newDomain = processDataFrameExpression(exitNode, new Map(inputDomain), dfg);
+				node = exitNode;
+			}
+		}
+		if(cfinfo.exitPoints.includes(vertex.id)) {
+			finalDomain = newDomain;
+		}
+		const visitedCount = visited.get(vertex.id) ?? 0;
+		visited.set(vertex.id, visitedCount + 1);
+
+		if(visitedCount >= WideningThreshold) {
+			newDomain = wideningDataFrameStates(oldDomain, newDomain);
+		}
+		if(node !== undefined) {
+			node.info.dataFrame ??= {};
+			node.info.dataFrame.domain = new Map(newDomain);
+		}
+		if(!equalDataFrameState(oldDomain, newDomain)) {
+			return getSuccessorVertices(cfg, vertex.id, dfg);
+		}
+		return getSuccessorVertices(cfg, vertex.id, dfg).filter(successor => !visited.has(successor.id));
+	};
+	const cfg = flipCfg(cfinfo.graph);
+	const entryPoints = cfinfo.entryPoints
+		.map(id => cfg.vertices().get(id))
+		.filter(vertex => vertex !== undefined);
+
+	foldCfg(cfg, entryPoints, visitor);
+	return finalDomain;
+}
+
+export function flipCfg(cfg: ControlFlowGraph): ControlFlowGraph {
+	const flippedCfg = new ControlFlowGraph();
+
+	for(const [id, vertex] of cfg.vertices()) {
+		flippedCfg.addVertex(vertex, cfg.rootVertexIds().has(id));
+	}
+	for(const [to, edges] of cfg.edges()) {
+		for(const [from, edge] of edges) {
+			flippedCfg.addEdge(from, to, edge);
+		}
+	}
+	return flippedCfg;
+}
+
+function foldCfg(
+	cfg: ControlFlowGraph,
+	vertices: CfgVertex[],
+	visitor: (cfg: ControlFlowGraph, vertex: CfgVertex) => CfgVertex[]
+): void {
+	for(const vertex of vertices) {
+		const successors = visitor(cfg, vertex);
+		foldCfg(cfg, successors, visitor);
+	}
+}
+
+function isRConstant(
+	node: RNode<ParentInformation>
+): node is RConstant<ParentInformation> {
+	return node.type === RType.String || node.type === RType.Number || node.type === RType.Logical;
+}
+
+function isRSingleNode(
+	node: RNode<ParentInformation>
+): node is RSingleNode<ParentInformation> {
+	return isRConstant(node) || node.type === RType.Symbol || node.type === RType.Break || node.type === RType.Next || node.type === RType.Comment || node.type === RType.LineDirective;
+}
```

Comment:
I think we should move these functions to a more general position - maybe alongside the model?